### ISSUE TITLE
Numerical issues with Stochastic two-body problem

### ISSUE THREAD

**Issue by matteoettam09** [SciML/DifferentialEquations.jl#857]
Hello, I am currently building on this work:

https://aip.scitation.org/doi/pdf/10.1063/1.4906908

and I am starting by replicating their results, and actually by having an implementation with all zeros in the diffusion term so that the energy is conserved so that I have a benchmark to start from. This is already proving a challenging step, for me, as we are getting diffusive behaviors solely due to numerical errors.

This
```julia 
sol = solve(prob_sde_2bp, EM(), dt=0.00001)
````
leads to
![stochastic_energy](https://user-images.githubusercontent.com/33622899/160902143-0643d4ca-1111-43ad-97bf-755ed8ea8a6b.png)

and this
```julia 
sol = solve(prob_sde_2bp, LambaEM(), tol=1e-9, reltol=1e-9, abstol=1e-9)
````
leads to
![stochastic_energy](https://user-images.githubusercontent.com/33622899/160901076-90097d09-5e4a-4a69-be22-96854f97265c.png)

here is the entire script:
```julia
μ = 10.0
u0 = [1.0,0.0,0.01,1.1]
tspan = (0.0,5.0)
function Sharma2bp(du,u,p,t)
  du[1] = u[3]
  du[2] = u[4]
  du[3] = u[1]*u[4]^2 - μ/u[1]^2
  du[4] = -2*u[3]*u[4]/u[1]
end

prob = ODEProblem(Sharma2bp,u0,tspan)
sol = solve(prob, Tsit5(), reltol=1e-8, abstol=1e-8)

r = sol[1, :]
theta = sol[2, :]
V = sqrt.(sol[3,:].^2 + (sol[4,:].*sol[1,:]).^2)
energy = V.^2 / 2 - μ./r
plot(energy)
savefig("deterministic_energy.png")

x = r.*cos.(theta)
y = r.*sin.(theta)
plot(x, y)
savefig("deterministic_traj.png")

function σ_2bp(du,u,p,t)
  du[1,1] = 0.0
  du[1,2] = 0.0
  du[2,1] = 0.0
  du[2,2] = 0.0
  du[3,1] = 0.0 # 0.0121*u[1]
  du[3,2] = 0.0
  du[4,1] = 0.0
  du[4,2] = 0.0 # 0.00022/u[1]
end

prob_sde_2bp = SDEProblem(Sharma2bp,σ_2bp, u0, tspan, noise_rate_prototype=zeros(4,2))
sol = ....
```

I have noticed that a different initial condition like
```julia
u0 = [1.0,0.0,0.01,1.1]
```
gives good results, so the problem is probably related to stiffness(?)

In the paper, they talk about an RK method of order 2..any idea how to get a good implementation with diffeq?

Thanks!

**Comment by tom-plaa** [comment]
Not to speak on behalf of the devs, but what do you mean by diffusive behaviours? Just looking at your plots, they seem pretty much deterministic to me.

**Comment by matteoettam09** [comment]
By diffusive I mean variation of quantities that should analytically be conserved, such as the total energy in the context of the two-body problem: the variation of y1 in both plots is solely due to numerical error: we would like to arrive at a situation in which the variation here is zero so that we can "turn-on" the diffusion terms and assess their effect on the variation of, e.g., energy. Thank you.

**Comment by matteoettam09** [comment]
Thank you Tom. As we are not experienced with sde numerical integration schemes, I didn't think that EM reduces to Euler when we set sigma to zero. What you are proposing makes sense, but as we already have experience with deterministic systems in celestial mechanics, we wanted to use already-existing numerical schemes in Julia for sdes that are characterized by non-diagonal noise. I am now thinking that people in celestial mechanics usually work with symplectic integration schemes, leveraging the conservation of the Hamiltonian for deterministic systems. Are you aware/are there in Julia numerical schemes that generalize this concept of symplectic systems for Hamiltonian systems with a stochastic perturbation?

Thank you

**Comment by tom-plaa** [comment]
The solving schemes I was mentioning are already built in into julia; they are in the documentation pages I included in the previous comment. Regarding your comment on sympletic methods, I am not familiar with them but I did find the following paragraph (cited from the documentation at https://diffeq.sciml.ai/stable/solvers/sde_solve/):

> If the noise term is not too large, for stiff problems with diagonal noise, ImplicitRKMil is the most efficient method and can choose between Ito and Stratonovich. For each of the theta methods, the parameter theta can be chosen. The default is theta=1/2 which will not dampen numerical oscillations and thus is symmetric (and almost symplectic) and will lead to less error when noise is sufficiently small. However, theta=1/2 is not L-stable in the drift term, and thus one can receive more stability (L-stability in the drift term) with theta=1, but with a tradeoff of error efficiency in the low noise case. In addition, the option symplectic=true will turns these methods into an implicit Midpoint extension which is symplectic in distribution but has an accuracy tradeoff.

**Comment by matteoettam09** [comment]
Great, thank you very much! I missed the discussion about symplectic methods in the documentation! Will take a look, thank you, the issue is closed for me.

Best,
Matteo

**Comment by ChrisRackauckas** [comment]
The ODE symplectic schemes are all here:

https://diffeq.sciml.ai/stable/solvers/dynamical_solve/

Indeed, the standard schemes are not symplectic and so they will have a numerical drift in long simulations. There are symplectic methods we are going to implement:

https://github.com/SciML/StochasticDiffEq.jl/issues/181

but that'll take some time.

However, what you're looking at is a bit more sinister than just that. Look at the paper:

>Our results are supported by numerical simulations which are obtained using a specific stochastic Runge-Kutta introduced by Kasdin and Stankievech in Ref. 14.

Ho boy, I've talked about Kasdin's method before:

https://discourse.julialang.org/t/has-anyone-used-kasdins-runge-kutta-sde-algorithm/68198/2?u=chrisrackauckas

In summary, it makes some well-known errors and so correct SDE integrators will not recreate its results.

**Comment by tom-plaa** [comment]
> https://discourse.julialang.org/t/has-anyone-used-kasdins-runge-kutta-sde-algorithm/68198/2?u=chrisrackauckas

Very interesting discussion
